skiko/src/commonMain/cpp/common/node/RenderNode.cpp

```diff
@@ -0,0 +1,337 @@
+    this->spotShadowColor = spotShadowColor;
+}
+
+void RenderNode::setRotationX(float rotationX) {
+    this->rotationX = rotationX;
+    this->matrixDirty = true;
+}
+
+void RenderNode::setRotationY(float rotationY) {
+    this->rotationY = rotationY;
+    this->matrixDirty = true;
+}
+
+void RenderNode::setRotationZ(float rotationZ) {
+    this->rotationZ = rotationZ;
+    this->matrixDirty = true;
+}
+
+float RenderNode::getCameraDistance() const {
+    return this->transformCamera.fLocation.z / 72.0f;
+}
+
+void RenderNode::setCameraDistance(float cameraDistance) {
+    this->setCameraLocation(0.0f, 0.0f, cameraDistance);
+    this->matrixDirty = true;
+}
+
+void RenderNode::setClipRect(const std::optional<SkRect>& clipRect) {
+    this->clipRect = clipRect;
+    this->clipRRect.reset();
+    this->clipPath.reset();
+}
+
+void RenderNode::setClipRRect(const std::optional<SkRRect>& clipRRect) {
+    this->clipRect.reset();
+    this->clipRRect = clipRRect;
+    this->clipPath.reset();
+}
+
+void RenderNode::setClipPath(const std::optional<SkPath>& clipPath) {
+    this->clipRect.reset();
+    this->clipRRect.reset();
+    this->clipPath = clipPath;
+}
+
+void RenderNode::setClip(bool clip) {
+    this->clip = clip;
+}
+
+const SkMatrix& RenderNode::getMatrix() {
+    this->updateMatrix();
+    return this->transformMatrix;
+}
+
+SkCanvas *RenderNode::beginRecording() {
+    bool measureDrawBounds = !clip || shadowElevation > 0.0f;
+    const SkRect& bounds = measureDrawBounds ? PICTURE_BOUNDS : SkRect::MakeWH(this->bounds.width(), this->bounds.height());
+    SkBBHFactory* bbhFactory = measureDrawBounds ? this->bbhFactory : nullptr;
+    return this->recorder.beginRecording(bounds, bbhFactory);
+}
+
+void RenderNode::endRecording() {
+    this->contentCache = this->recorder.finishRecordingAsDrawable();
+}
+
+void RenderNode::drawInto(SkCanvas* canvas) {
+    canvas->drawDrawable(this);
+}
+
+void RenderNode::onDraw(SkCanvas* canvas) {
+    this->updateMatrix();
+
+    canvas->translate(this->bounds.left(), this->bounds.top());
+    if (!this->matrixIdentity) {
+        canvas->concat(this->transformMatrix);
+    }
+
+    if (this->shadowElevation > 0) {
+        auto lightGeometry = this->context->getLightGeometry();
+        auto lightInfo = this->context->getLightInfo();
+        this->drawShadow(canvas, lightGeometry, lightInfo);
+    }
+
+    if (this->clip) {
+        canvas->save();
+        if (this->clipRect) {
+            canvas->clipRect(*this->clipRect);
+        } else if (this->clipRRect) {
+            canvas->clipRRect(*this->clipRRect);
+        } else if (this->clipPath) {
+            canvas->clipPath(*this->clipPath);
+        } else {
+            canvas->clipRect(SkRect::MakeWH(this->bounds.width(), this->bounds.height()));
+        }
+    }
+
+    if (this->layerPaint) {
+        canvas->saveLayer(SkRect::MakeWH(this->bounds.width(), this->bounds.height()), &*this->layerPaint);
```

Comment:
Where are we  doing `canvas->restore()`?

Reply:
Skia maintains this. `onDraw` is wrapped to auto restoring primitives.

Decided to remove duplication on our side after looking at skia's sources